text/0080-datetime-extension.md

```diff
@@ -0,0 +1,490 @@
+
+This RFC proposes supporting two new extension types: `datetime`, which represents a particular instant of time, up to second accuracy, and `duration` which represents a duration of time.
+To construct and manipulate these types we will provide the functions listed below.
+All of this functionality will be hidden behind a `datetime` feature flag (analogously to the current decimal and IP extensions), allowing users to opt-out if they do not want this functionality.
+
+- `datetime(s)` constructs a datetime object. Like with our other constructors, strict validation requires `s` to be a string literal, although evaluation/authorization support any string-typed expression. The string must have the form `"YYYY-MM-DDThh:mm:ssZ"` (UTC), or `"YYYY-MM-DDThh:mm:ss(+/-)hhmm"` (with Time Zone Offset).
+  - This format is a subset of ISO 8601.
+  - Internally, the returned `datetime` is always stored as UTC.
+- `d.offset(du)` returns a new date with time offset by duration `du`.
+  - This allows for a "lightweight" implementation of local time.
+- `d.isBetween(d1, d2)` checks whether date `d` is between dates `d1` and `d2` (inclusive).
+- `d.isBefore(d1)` checks whether date `d` is before date `d1` (non-inclusive).
+- `d.since(d1)` computes the difference between two `datetime` objects and returns a `duration` object.
+- `d.year()`, `d.month()`, `d.day()`, `d.hour()`, `d.minute()`, `d.second()` returns the relevant component of the date.
+- `d.dayOfWeek()` returns 1 when the day is Sunday, and 7 when the day is Saturday.
+
+Internally, the datetime object will have a structure such as:
+
+```cedarschema
+{
+  year: long,
+  month: long,
+  day: long,
+  hours: long,
+  minutes: long,
+  seconds: long,
+}
+```
+
+The datetime parsing function will be responsible for ensuring validity of the structure (e.g., that `month` is between 1 and 12, see below for full details).
+
+The `d.since(d1)` function returns a `duration` object.
+
+- `duration(n, unit)` returns a `duration` object. Strict validation requires `unit` to be a string literal of one of the following values: "seconds", "minutes", "hours", "days".
+- `du.toSeconds()`, `du.toMinutes()`, `du.toHours()`, `du.toDays()` returns a `long` indicating the number of units represented by this duration, truncated as appropriate. _e.g._
+  - A `duration` representing 119 seconds, returns `1` when `toMinutes()` is called.
+  - A `duration` representing 10 minutes, returns `0` when `toHours()` is called.
+
+Internally, the `duration` object is a tagged `long` representing seconds. As a record:
+
+```cedarschema
+{
+  seconds: long,
+}
+```
+
+### Constraints on dates
+
+Here are the full validity constraints on datetime values:
+
+- `year` must be between 1900 and 2100
+- `month` must be between 1 and 12
+- `day` must be between 1 and...
+  - 30 for `month`=4,6,9,11
+  - 31 for `month`=1,3,5,7,8,10,12
+  - 28 for `month`=2 if `year` is not divisible by 4
+  - 29 for `month`=2 if `year` is divisible by 4
```

Comment:
I think this should at least be `y % 400 == 0 || (y % 4 == 0 && (y % 100 != 0))`

Reply:
Correct, but the RFC puts a (I think? practical) constraint on the year to be between 1900 and 2100, in an attempt to simplify the quirks of the Gregorian calendar (e.g. leap years).

Reply:
I missed that. (But I also don't like it 😄)

Reply:
Ha! I didn't like it either, but, it simplified things! :)